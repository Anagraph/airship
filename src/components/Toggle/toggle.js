import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { colors } from '../../constants';

const Check = () => {
  return (
    <svg width="10" height="8" xmlns="http://www.w3.org/2000/svg">
      <path
        d="M3.315 7.858L.133 4.441a.506.506 0 0 1 0-.683l.643-.684a.437.437 0 0 1 .642 0l2.219 2.393L8.58.141a.437.437 0 0 1 .643 0l.643.683a.504.504 0 0 1 0 .683l-5.91 6.35a.437.437 0 0 1-.642 0"
        fill="#FFF"
        fillRule="evenodd"
      />
    </svg>
  );
};

const StyledToggle = styled.div`
  display: inline-block;
  position: relative;

  &,
  &:after,
  &:before,
  & *,
  & *:after,
  & *:before {
    box-sizing: border-box;
  }

  svg {
    opacity: 0;
    position: absolute;
    left: 4px;
    top: 4px;
    transition: opacity 0.2s ease;
    pointer-events: none;
  }

  input:checked ~ svg {
    opacity: 1;
  }

  input {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }

  input + label {
    width: 32px;
    height: 16px;
    outline: 0;
    display: block;
    position: relative;
    cursor: pointer;
    user-select: none;
    background: ${colors.ui04};
    border-radius: 40px;
    transition: all 0.4s ease;
  }

  input + label::after {
    position: relative;
    display: block;
    content: '';
    width: 12px;
    height: 12px;
    top: 2px;
  }

  input + label::after {
    border-radius: 50%;
    background: #fff;
    transition: all 0.2s ease;
    transform: translateX(2px);
  }

  label:hover {
    background: ${colors.brand03};
  }

  input[disabled] + label {
    cursor: not-allowed;
    background: ${colors.ui02};
    box-shadow: inset 0 0 0 1px ${colors.ui03};
  }

  input[disabled] + label:after {
    background: ${colors.ui04};
  }

  input:checked + label:hover,
  input:checked + label {
    background: ${colors.brand01};
  }

  input:checked + label::after {
    transform: translateX(18px);
  }
`;
StyledToggle.displayName = 'StyledToggle';

const StyledControl = styled.div`
  display: flex;
  align-items: center;
`;

const StyledLabel = styled.div`
  margin-left: 8px;
`;

class Toggle extends Component {
  state = {
    checked: !!this.props.checked
  };

  componentWillReceiveProps(nextProps) {
    if (this.state.checked !== nextProps.checked) {
      this.setState({ checked: nextProps.checked });
    }
  }

  clickHandler = (e) => {
    const { onChange, disabled } = this.props;
    !disabled &&
      this.setState(
        (state) => {
          return {
            ...state,
            checked: !state.checked
          };
        },
        () => {
          onChange && onChange(this.state);
        }
      );
  };

  render() {
    const { htmlFor, disabled, children } = this.props;
    const { checked } = this.state;

    return (
      <StyledControl>
        <StyledToggle>
          <input
            id={htmlFor}
            type="checkbox"
            disabled={disabled ? 'disabled' : null}
            checked={checked}
            onChange={this.clickHandler}
          />
          <label htmlFor={htmlFor} />
          <Check />
        </StyledToggle>
        {children ? <StyledLabel>{children}</StyledLabel> : null}
      </StyledControl>
    );
  }
}

Toggle.propTypes = {
  htmlFor: PropTypes.string.isRequired,
  onChange: PropTypes.func
};

export default Toggle;
