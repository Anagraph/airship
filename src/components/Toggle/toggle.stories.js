import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import { action } from '@storybook/addon-actions';
import Toggle from './toggle';
import Text from '../Typography/text';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  brand01: '#000',
  brand03: '#777',
  ui04: '#555',
  ui01: '#FABADA',
};

storiesOf('Toggle', module)
  .add('Default', () => (
    <div>
      <h3 className="header">Default</h3>
      <Toggle htmlFor="default">
        <Text>Hello there</Text>
      </Toggle>

      <h3 className="header">With label</h3>
      <Toggle htmlFor="with-label">
        <Text>General Kenobi</Text>
      </Toggle>

      <h3 className="header">Checked</h3>
      <Toggle htmlFor="checked" checked />

      <h3 className="header">Disabled</h3>
      <Toggle htmlFor="disabled" disabled />

      <h3 className="header">With event</h3>
      <Toggle htmlFor="with-event" onChange={action('onChange')}>
        <Text>Watto</Text>
      </Toggle>
    </div>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <div>
        <h3 className="header">Default</h3>
        <Toggle htmlFor="default">
          <Text>Hello there</Text>
        </Toggle>

        <h3 className="header">With label</h3>
        <Toggle htmlFor="with-label">
          <Text>General Kenobi</Text>
        </Toggle>

        <h3 className="header">Checked</h3>
        <Toggle htmlFor="checked" checked />

        <h3 className="header">Disabled</h3>
        <Toggle htmlFor="disabled" disabled />
      </div>
    </ThemeProvider>
  ));
