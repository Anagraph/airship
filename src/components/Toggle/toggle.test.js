import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import Toggle from './toggle';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Toggle htmlFor="foo" />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders with checked param', () => {
    const component = renderer.create(<Toggle htmlFor="foo" checked />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call callback when changed', () => {
    const spy = jest.fn();
    const component = shallow(<Toggle htmlFor="foo" onChange={spy} />);

    const input = component.find('input').first();
    input.simulate('change');

    expect(spy).toHaveBeenCalled();
  });

  it('sets \'checked\' state correcly when prop is changed', () => {
    const wrapper = mount(<Toggle htmlFor="foo" />);

    wrapper.setProps({ checked: true });

    expect(wrapper.state().checked).toBe(true);
  });
});
