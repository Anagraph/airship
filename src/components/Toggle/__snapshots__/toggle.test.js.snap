// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders with checked param 1`] = `
.c1 {
  display: inline-block;
  position: relative;
}

.c1,
.c1:after,
.c1:before,
.c1 *,
.c1 *:after,
.c1 *:before {
  box-sizing: border-box;
}

.c1 svg {
  opacity: 0;
  position: absolute;
  left: 4px;
  top: 4px;
  -webkit-transition: opacity 0.2s ease;
  transition: opacity 0.2s ease;
  pointer-events: none;
}

.c1 input:checked ~ svg {
  opacity: 1;
}

.c1 input {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 input + label {
  width: 32px;
  height: 16px;
  outline: 0;
  display: block;
  position: relative;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #D1D5D7;
  border-radius: 40px;
  -webkit-transition: all 0.4s ease;
  transition: all 0.4s ease;
}

.c1 input + label::after {
  position: relative;
  display: block;
  content: '';
  width: 12px;
  height: 12px;
  top: 2px;
}

.c1 input + label::after {
  border-radius: 50%;
  background: #fff;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  -webkit-transform: translateX(2px);
  -ms-transform: translateX(2px);
  transform: translateX(2px);
}

.c1 label:hover {
  background: #47DB99;
}

.c1 input[disabled] + label {
  cursor: not-allowed;
  background: #F5F5F5;
  box-shadow: inset 0 0 0 1px #E2E6E3;
}

.c1 input[disabled] + label:after {
  background: #D1D5D7;
}

.c1 input:checked + label:hover,
.c1 input:checked + label {
  background: #1785FB;
}

.c1 input:checked + label::after {
  -webkit-transform: translateX(18px);
  -ms-transform: translateX(18px);
  transform: translateX(18px);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  margin-left: 8px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <input
      checked={true}
      disabled={null}
      id="foo"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="foo"
    />
    <svg
      height="8"
      width="10"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.315 7.858L.133 4.441a.506.506 0 0 1 0-.683l.643-.684a.437.437 0 0 1 .642 0l2.219 2.393L8.58.141a.437.437 0 0 1 .643 0l.643.683a.504.504 0 0 1 0 .683l-5.91 6.35a.437.437 0 0 1-.642 0"
        fill="#FFF"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="c2"
  />
</div>
`;

exports[`render renders without crashing 1`] = `
.c1 {
  display: inline-block;
  position: relative;
}

.c1,
.c1:after,
.c1:before,
.c1 *,
.c1 *:after,
.c1 *:before {
  box-sizing: border-box;
}

.c1 svg {
  opacity: 0;
  position: absolute;
  left: 4px;
  top: 4px;
  -webkit-transition: opacity 0.2s ease;
  transition: opacity 0.2s ease;
  pointer-events: none;
}

.c1 input:checked ~ svg {
  opacity: 1;
}

.c1 input {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 input + label {
  width: 32px;
  height: 16px;
  outline: 0;
  display: block;
  position: relative;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #D1D5D7;
  border-radius: 40px;
  -webkit-transition: all 0.4s ease;
  transition: all 0.4s ease;
}

.c1 input + label::after {
  position: relative;
  display: block;
  content: '';
  width: 12px;
  height: 12px;
  top: 2px;
}

.c1 input + label::after {
  border-radius: 50%;
  background: #fff;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  -webkit-transform: translateX(2px);
  -ms-transform: translateX(2px);
  transform: translateX(2px);
}

.c1 label:hover {
  background: #47DB99;
}

.c1 input[disabled] + label {
  cursor: not-allowed;
  background: #F5F5F5;
  box-shadow: inset 0 0 0 1px #E2E6E3;
}

.c1 input[disabled] + label:after {
  background: #D1D5D7;
}

.c1 input:checked + label:hover,
.c1 input:checked + label {
  background: #1785FB;
}

.c1 input:checked + label::after {
  -webkit-transform: translateX(18px);
  -ms-transform: translateX(18px);
  transform: translateX(18px);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  margin-left: 8px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <input
      checked={false}
      disabled={null}
      id="foo"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="foo"
    />
    <svg
      height="8"
      width="10"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.315 7.858L.133 4.441a.506.506 0 0 1 0-.683l.643-.684a.437.437 0 0 1 .642 0l2.219 2.393L8.58.141a.437.437 0 0 1 .643 0l.643.683a.504.504 0 0 1 0 .683l-5.91 6.35a.437.437 0 0 1-.642 0"
        fill="#FFF"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="c2"
  />
</div>
`;
